src/shims/unix/mem.rs

```diff
@@ -0,0 +1,240 @@
+//! This is an incomplete implementation of mmap/mremap/munmap which is restricted in order to be
+//! implementable on top of the existing memory system. The point of these function as-written is
+//! to allow memory allocators written entirely in Rust to be executed by Miri. This implementation
+//! does not support other uses of mmap such as file mappings.
+//!
+//! mmap/mremap/munmap behave a lot like alloc/realloc/dealloc, and for simple use they are exactly
+//! equivalent. But the memory-mapping API provides more control. For example:
+//!
+//! * It is possible to munmap a single page in the middle of a mapped region. We do not have a way
+//! to express non-contiguous allocations.
+//!
+//! * With MAP_FIXED it is possible to call mmap multiple times, but create a single contiguous
+//! range of mapped virtual addresses. A memory allocator can then choose to carve this up into
+//! allocations in arbitrary ways.
+//!
+//! For now, none of this is possible with the Miri implementation.
+
+use crate::*;
+use rustc_target::abi::{Align, Size};
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriInterpCx<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
+    fn mmap(
+        &mut self,
+        addr: &OpTy<'tcx, Provenance>,
```

Comment:
I think we should check that this is null, and error out as unsupported otherwise. The libc behaviour documented for non-null address is really weird

> If addr is not NULL, then the
       kernel takes it as a hint about where to place the mapping; on
       Linux, the kernel will pick a nearby page boundary (but always
       above or equal to the value specified by
       /proc/sys/vm/mmap_min_addr) and attempt to create the mapping
       there.  If another mapping already exists there, the kernel picks
       a new address that may or may not depend on the hint.

and we're not doing that, so we should not act as if we do support that, unless that is what crates need, (then we should test and document it).

Reply:
I think we support this as much as any normal `mmap` implementation does in my experience: We completely ignore it unless you set `MAP_FIXED`.

I'm willing to set up an error if you insist, but the man page doesn't say an implementation has to, and it would be odd for Miri to not pedantically follow the man pages.

Reply:
hmm... I agree that this is what the spec says, it still seems like a bug to pass a non-null pointer unless `MAP_FIXED` is used. Like... the user may expect this pointer to be used $somehow, even if just due to not having read the docs right.

Anyway, seems like an orthogonal discussion, not relevant for this PR specifically